File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        int sum = 0, cnt = 0;
        unordered_map<int, int> preMap;
        preMap[0] = 1;

        for(int i=0;i<n;i++) {
            sum += nums[i];
            int req = sum - k;

            if(preMap.find(req) != preMap.end()) {
                cnt += preMap[req];
            }
            
            preMap[sum]++;
        }

        return cnt;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n = nums.length;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);

        int sum = 0, cnt = 0;

        for(int i=0;i<n;i++) {
            sum += nums[i];
            int rem = sum - k;

            if(map.containsKey(rem)) {
                cnt += map.get(rem);
            }

            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }

        return cnt;
    }
}